portal/components/react-admin/base/resources/donate-device.js

```diff
@@ -279,11 +319,142 @@ export const DonateDeviceRequestEdit = (props) => {
         const response = await sendSMS(message, templateId, data.phone_number);
         if (response?.success) notify(response.success, "info");
         else if (response?.error) notify(response.error, "warning");
-        redirect("list", props.basePath, data.id, data);
+      }
+      redirect("list", props.basePath, data.id, data);
+    }
+  };
+
+  const validateForm = async (values) => {
+    const errors = {};
+    
+    if (values.delivery_status && values.delivery_status == "delivered-child") {
+      if (!values.device_verification_record?.otp) {
+        errors.device_verification_record = {otp:"The Otp is required"};
+      }
+      if (!values.device_verification_record?.verifier_phone_number) {
+        errors.device_verification_record = {...errors.device_verification_record,verifier_phone_number:"Verifier's phone number is required"};
+      } else if(!values.device_verification_record.verifier_phone_number.match('[0-9]{10}')) {
+        errors.device_verification_record = {...errors.device_verification_record,verifier_phone_number:"Enter a valid 10 digit mobile number."};
+      }
+    }
+
+    return errors;
+  };
+
+  const uuidv4 = () => {
+    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(
+      /[xy]/g,
+      function (c) {
+        var r = (Math.random() * 16) | 0,
+          v = c == "x" ? r : (r & 0x3) | 0x8;
+        return v.toString(16);
+      }
+    );
+  };
+
+  const filtered = (raw, allowed, except = "only") => {
+    return Object.keys(raw)
+      .filter((key) =>
+        except == "except" ? !allowed.includes(key) : allowed.includes(key)
+      )
+      .reduce((obj, key) => {
+        obj[key] = raw[key];
+        return obj;
+      }, {});
+  };
+
+  const save = useCallback(
+    async (values) => {
+      try {
+        const recordData = filtered(values, ['device_verification_record'], "except");
```

Comment:
Can we turn the OTP functions into utility functions that we can use across all components?